Document mixed `PackageReference` and `packages.config` projects (#628)
microsoft/react-native-windows-samples: docs/managing-cpp-deps.md

---
id: managing-cpp-deps
title: Managing C++ dependencies
---

> Details to consider when consuming a Visual C++ project as a source dependency of your app.

Reviewer: A major part of the RNW demographic will be familiar with Visual Studio project structure. From what I've observed, others have expertise in JS, and rely on tools like "autolinking" to manage dependencies for them.

That audience may not know, that under the hood, their npm package dependency brings in a vcxproj. The term the community would recognize for these npm dependencies bringing in native code is "community modules". Since that has the potential to be such a common case, could we call them out by name throughout the document?

E.g.

```suggestion
Details to consider when consuming community modules or other Visual C++ projects.
```

### Details

Applications with native code, either written on `C#` or `C++`, may add source dependencies on native Visual C++ (`.vcxproj`) projects.\
Meaning, such dependencies will be built as part of the application.

Starting with version `0.68`, React Native for Windows apps use the [`PackageReference`](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files) restore project style for native `C++` NuGet dependencies.\
The main change consists in NuGet packages being directly loaded from the user account's `globalPackagesFolder` cache instead of also copying them into a local folder relative to the Visual Studio Solution location (see [`repositoryPath`](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file)).

> This may conflict with Node/Visual C++ dependencies that use the more common [`packages.config`](https://docs.microsoft.com/en-us/nuget/reference/packages-config) project style (i.e. [React Native Picker](https://github.com/react-native-picker/picker#react-native-pickerpicker)).

Reviewer: There are many community modules, so the reader might be using a different one, and not be familiar with Picker. 

```suggestion
This may conflict with Visual C++ dependencies that use the more common [`packages.config`](https://docs.microsoft.com/en-us/nuget/reference/packages-config) project style, including community modules generated targeting versions before 0.68.
```

Reviewer: Same feedback for below, for anywhere we reference anything specific to Picker.

## Updating your app

### Add the React Native Windows C++ Dependency (i.e. RN Picker)

```
yarn add @react-native-picker/picker
```

> ### Add a `packages.config` file to your app's project directory

Reviewer: Historically, autolinking would manage this file. Is that still the case?

Reviewer: Er, I might be mixing this up. I remember there was some differences of NuGet restore behavior for transitive dependencies on packages.config that @jonthysell ran into before.

Reviewer 2: Wait, is this just as simple as adding a packages.config to that app's project directory?

Is there any reason we're not just putting/modifying a file there ourselves during autolinking? Or just manually calling nuget restore on the lib project so the files get put in the right place?

Author: Calling NuGet restore doesn't work once PackageReference is detected.

I did not know we could have Autolink set this up for us.
We could add such functionality and then remove this instruction.

---
id: managing-cpp-deps
title: Managing C++ dependencies
---

Details to consider when consuming community modules or other Visual C++ projects.

### Details

Applications with native code, either written on `C#` or `C++`, may add source dependencies on native Visual C++ (`.vcxproj`) projects.\
Meaning, such dependencies will be built as part of the application.

> Starting with version `0.68`, React Native for Windows apps use the [`PackageReference`](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files) restore project style for native `C++` NuGet dependencies.\
> The main change consists in NuGet packages being directly loaded from the user account's `globalPackagesFolder` cache instead of also copying them into a local folder relative to the Visual Studio Solution location (see [`repositoryPath`](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file)).
> 
> This may conflict with Visual C++ dependencies that use the more common [`packages.config`](https://docs.microsoft.com/en-us/nuget/reference/packages-config) project style, including community modules generated targeting versions before 0.68.

Reviewer 3: ```suggestion
Starting with version `0.68`, React Native for Windows apps use the [`PackageReference`](https://docs.microsoft.com/nuget/consume-packages/package-references-in-project-files) restore project style for native `C++` NuGet dependencies.\
The main change consists of NuGet packages being placed directly in the user account's `globalPackagesFolder` cache instead of also copying them into a local folder relative to the Visual Studio Solution location (see [`repositoryPath`](https://docs.microsoft.com/nuget/reference/nuget-config-file)).

This may conflict with C++ dependencies that use the more common [`packages.config`](https://docs.microsoft.com/nuget/reference/packages-config) project style, including community modules generated targeting versions older than 0.68.
```

```
yarn add @react-native-picker/picker
```

### Add a `packages.config` file to your app's project directory

The React Native Windows build system will try to build the dependency from source.\
Because it uses the `packages.config` restore style, it will most likely expect its own NuGet dependencies to be restored at `$(SolutionDir)packages\`.

#### See [`node_modules\@react-native-picker\picker\windows\ReactNativePicker\ReactNativePicker.vcxproj`](https://github.com/react-native-picker/picker/blob/v2.2.1/windows/ReactNativePicker/ReactNativePicker.vcxproj#L156)

```xml title="ReactNativePicker.vcxproj"
<Import Project="$(SolutionDir)\packages\Microsoft.Windows.CppWinRT.2.0.210312.4\build\native\Microsoft.Windows.CppWinRT.targets" Condition="Exists('$(SolutionDir)\packages\Microsoft.Windows.CppWinRT.2.0.210312.4\build\native\Microsoft.Windows.CppWinRT.targets')" />
<Import Project="$(SolutionDir)\packages\$(WinUIPackageName).$(WinUIPackageVersion)\build\native\$(WinUIPackageName).targets" Condition="Exists('$(SolutionDir)\packages\$(WinUIPackageName).$(WinUIPackageVersion)\build\native\$(WinUIPackageName).targets')" />
```

> #### `windows\<C++ or C# app directory>\packages.config`

Reviewer 3: if the app is C#, would it ever use packages.config?

Author: Not directly, but a C# project could depend on a C++ project that uses `packages.config`.

For instance, a MSRN C# app could depend on RN Picker, couldn't it?
